CONTRIBUTING.md

```diff
@@ -1,73 +1,76 @@
-- Sometimes, we have convenience functions, such as `ReferenceSegment::operator[]`, or workarounds, such as performing multiple stable sorts instead of a single one. Because these might negatively affect the performance, the user should be warned if a query causes one of these slow paths to be chosen. For this, we have the PerformanceWarning() macro defined in `performance_warning.hpp`.
-
-## Documentation
-- Most documentation should happen in the code or in the beginning of the header file
-- More complex documentation, such as an explanation of an algorithm that profits from images, can be put in the Wiki. Please make sure to link the Wiki page in the code - otherwise, no one will find it.
-
-# Review
-
-**Things to look for:**
-
-	- Guidelines (see above)
-	- Is the copy constructor deleted where it makes sense?
-	- Is the destructor virtual for base classes?
-	- Are unnecessary copies of heavy elements made? (prefer vector& over vector, but not int& over int)
-	- Did the author update documentation and dependencies (Wiki, README.md, DEPENDENCIES.md, Dockerfile, install_dependencies.sh, Jenkinsfile)
+# The Four Commandments
+1. Thy code (or the beginning of the header file) shalt be the primary method of documentation. Comments should be used to explain higher level concepts, but primarily, the code should be explaining itself. Part of this is to choose concise, descriptive names.
+1. Thou shalt program defensively. Use Assert wherever it makes sense, use DebugAssert in performance-critical parts (e.g., within hot loops). Also, we do not handle exceptions in Hyrise. If an invalid state is reached, we crash immediately. This makes debugging easier. An exception (hihi) to this is user-facing code where we handle, e.g., typos in SQL queries.
+1. Thou shalt reflect on existing code. Just as your code is not perfect, neither is the code that people wrote before you. Try to improve it as part of your PRs and do not hesitate to ask if anything is unclear. Chances are that it can be improved.
+1. Thou shalt properly test thy code. This includes unit *and* integration tests. Try to isolate parts that can be individually tested.
+
+# C++
+* Use automatic memory management (RAII, smart pointers). `new` and `malloc` are evil words.
+* Use `const` whenever possible. Consider variables, methods, pointers, and their pointees.
+* Reduce the size of your hpp files, both in terms of the number of lines and the code complexity. This keeps the compilation duration low.
+  * Code in hpp files is compiled for every cpp file that includes the header. As such, move code to cpp files where possible. This often includes templated classes, where it is sometimes possible to implement their code in cpp files.
+  * Anonymous namespaces are a good way to define local helper methods.
+  * Use forward declarations instead of full header includes wherever possible.
+* Loops
+  * Use range-based for loops when possible: `for (const auto& item : items) {...}`.
```

Comment:
Since I remarked this quite often the last weeks: also name structured bindings?

Reply:
Added

> Consider structured bindings: `const auto& [iterator, added] = unordered_map.emplace(...);`